src/python-bindings/classad3_impl.cpp

```diff
@@ -1 +1,401 @@
+
+			PyErr_SetString( PyExc_AssertionError, "_evaluate() passed invalid scope_str" );
+			return NULL;
+		}
+	}
+
+	classad::Value v;
+	ExprTree * flat = NULL;
+	if(! scope->Flatten( expr, v, flat )) {
+		delete( expr );
+		delete( scope );
+
+		// FIXME: See below.
+		PyErr_SetString( PyExc_ValueError, "Expression failed to flatten" );
+		return NULL;
+	}
+
+	std::string result;
+	classad::ClassAdUnParser unparser;
+	if( flat != NULL ) {
+		unparser.Unparse( result, flat );
+	} else {
+		unparser.Unparse( result, v );
+	}
+	return PyUnicode_FromString( result.c_str() );
+}
+
+
+bool
+EvaluateLooseExpr( classad::ExprTree * expr,
+                   classad::ClassAd * my, classad::ClassAd * target,
+                   classad::Value & value ) {
+	const classad::ClassAd * originalScope = expr->GetParentScope();
+	expr->SetParentScope(my);
+
+	bool rv = false;
+	if( target == my || target == NULL ) {
+		rv = expr->Evaluate( value );
+	} else {
+		classad::MatchClassAd matchAd( my, target );
+		rv = expr->Evaluate( value );
+		matchAd.RemoveLeftAd();
+		matchAd.RemoveRightAd();
+	}
+
+	expr->SetParentScope(originalScope);
+	return rv;
+}
+
+
+static PyObject *
+_evaluate( PyObject * /* self */, PyObject * args ) {
+	const char * expr_str = NULL;
+	const char * scope_str = NULL;
+	const char * target_str = NULL;
+
+	// 's': valid new ClassAds don't allow embedded NULs, I hope.
+	if(! PyArg_ParseTuple( args, "sss", & expr_str, & scope_str, & target_str )) {
+		// PyArg_ParseTuple() has already set an exception for us.
+		return NULL;
+	}
+
+
+	classad::ClassAdParser parser;
+
+	classad::ExprTree * expr = NULL;
+	if(! parser.ParseExpression( expr_str, expr, true )) {
+		PyErr_SetString( PyExc_AssertionError, "_evaluate() passed invalid expr_str" );
+		return NULL;
+	}
+	// FIXME: assign expr to a classad_shared_ptr now?
+
+	classad::ClassAd * scope = NULL;
+	if( scope_str[0] != '\0' ) {
+		scope = parser.ParseClassAd( scope_str, true );
+		if( scope == NULL ) {
+			delete( expr );  // FIXME?
+
+			PyErr_SetString( PyExc_AssertionError, "_evaluate() passed invalid scope_str" );
+			return NULL;
+		}
+	}
+
+	classad::ClassAd * target = NULL;
+	if( target_str[0] != '\0' ) {
+		target = parser.ParseClassAd( target_str, true );
+		if( target == NULL ) {
+			delete( expr ); // FIXME?
+			delete( scope ); // FIXME?
+
+			PyErr_SetString( PyExc_AssertionError, "_evaluate() passed invalid target_str" );
+			return NULL;
+		}
+	}
+
+
+	bool rv = false;
+	classad::Value value;
+	if( scope != NULL ) {
+		rv = EvaluateLooseExpr( expr, scope, target, value );
+	} else {
+		classad::EvalState state;
+		rv = expr->Evaluate( state, value );
+	}
+	if( PyErr_Occurred() ) { return NULL; }
+	if(! rv) {
+		delete( expr ); // FIXME?
+		delete( scope ); // FIXME?
+		delete( target ); // FIXME?
+
+		// This was ClassAdErrorValue in version 1.
+		// (Should we do the translation in the Python layer?)
+		PyErr_SetString( PyExc_ValueError, "Expression failed to evaluate" );
```

Comment:
I'd like to see a comparison of the traceback between raising the error here vs. raising it in `Expression.evaluate()`. I suspect if some function is going to know _why_ the expression failed to evaluate, it's probably this function, not the Python wrapper, so maybe here is better in case we decide to be more verbose about what failed?

Reply:
I tried and haven't been able to come up with an expression that's syntactically valid but which fails to evaluate.  But I faked it.

```>>> e.evaluate()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tlmiller/Work/condor/install/lib/python3/classad3/Expression.py", line 55, in evaluate
    answer_blob = _evaluate(self.expr_string, my_string, target_string)
ValueError: Expression failed to evaluate
```
vs
```>>> e.evaluate()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/tlmiller/Work/condor/install/lib/python3/classad3/Expression.py", line 56, in evaluate
    raise ValueError("Expression failed to evaluate")
ValueError: Expression failed to evaluate
```

Reply:
I'm happy with this as-is.